L52-objc-bazel-support.md

```diff
@@ -0,0 +1,124 @@
+* The Objective-C stubs are generated *and compiled* into native Bazel `objc_library`'s via `objc_proto_lib` and `objc_grpc_lib`. Details about these two custom rules are discussed in the upcoming sections.
+* Although "app-specific resources" depend on multiple libraries in the graph, users only need to add `objc_grpc_lib`'s and `objc_grpc_lib`'s they defined. This is because the dependency on gRPC-protoRPC and gRPC-ObjC are carried by `objc_grpc_lib`.
+* All the necessary external dependencies are loaded with `grpc_deps()` in `//bazel:grpc_deps.bzl` and are hidden from the users.
+* The app-specific code does need to depend on `//:gRPCCertificates`, which is essentially a `.pem` file.
+
+### Rules for Compiling `.proto` Files
+
+The gist of these custom rules is to run protobuf compiler and Objective-C plugin on provided `.proto` files with `ctx.action.run`. They are available from `@com_google_protobuf//:protoc` and `@com_github_grpc_grpc//:grpc_objective_c_plugin`.
+
+We use the native `proto_library` rule as a manager for `.proto` files (i.e. their package paths and dependencies). They wrap the proto files and are passed into `objc_proto_lib` and `objc_grpc_lib` as `srcs`. As a result of using `proto_library`, proto files must not reside in a upper-level directory to the `BUILD` file.
+
+Due to naming rules within the protobuf compiler, we first need to convert file names into camel case in order to correctly declare all the output files. File extensions are either `.pbobjc.h/.m` or `.pbrpc.h/.m`, give the pair of files that will be compiled (Messages or Services, respectively) as a boolean parameter.
+
+The ObjC plugin is one of the arguments. The precise arguments is formatted by the libary function `get_plugin_args` provided in `//bazel:protobuf.bzl`.
+
+If the proto files are depend on any of the "well known" proto files like `empty.proto` defined in `@com_google_protobuf//google/protobuf/`, this external rule `@com_google_protobuf//:well_known_protos` is needed in order to add those files into arguments.
+
+Outputs will be generated in directories managed by Bazel. They will be further splitted into two groups - one for `hdrs` and the other for `srcs`/`non_arc_srcs` - and then passed into a native `objc_library` rule for the final compiled library. Note that one `objc_proto/grpc_lib` might need to depend on others as well.
+
+### Example
+
+Suppose a very simple HelloWorldClient project with a single proto file `//HelloWorld/hello.proto` with both messages and services. Assume that it depends on one of the well known protos.
+
+Configure `WORKSPACE` as instructed in this [tutorial](https://docs.bazel.build/versions/master/tutorial/ios-app.html) and also load `grpc_deps`:
+```
+git_repository(
+    name = "com_github_grpc_grpc",
+    remote = "https://github.com/grpc/grpc.git",
+    branch = "master"
+)
+load("@com_github_grpc_grpc//bazel:grpc_deps.bzl", "grpc_deps")
+grpc_deps()
+```
+
+And here is the `BUILD` file:
```

Comment:
"The `BUILD` file of the example project can be written similar to the following snippet:"